File: XMLCheck/Handler/Xml2Parser.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using XMLCheck.Models;

namespace XMLCheck.Handler
{
    public class Xml2Parser : BaseXmlParser, IXmlParser
    {
        public string LoaiHoSo => "XML2";

        public object Parse(string xmlContent)
        {
            var result = new List<Xml2Model>();

            // 1. Load XML (Base xử lý BOM + Parse)
            if (!Load(xmlContent))
                return result;

            // 2. Check root: <CHITIEU_CHITIET_THUOC>
            if (!IsRoot("CHITIEU_CHITIET_THUOC"))
                return result;

            // 3. Lấy DSACH_CHI_TIET_THUOC
            var dsachNodes = Root!
                .Elements()
                .Where(e => e.Name.LocalName == "DSACH_CHI_TIET_THUOC");

            foreach (var dsach in dsachNodes)
            {
                // 4. Lấy từng CHI_TIET_THUOC
                var itemNodes = dsach
                    .Elements()
                    .Where(e => e.Name.LocalName == "CHI_TIET_THUOC");

                foreach (var n in itemNodes)
                {
                    // ===== Helper CHỈ đọc trong CHI_TIET_THUOC =====
                    string? S(string k) =>
                        n.Elements()
                         .FirstOrDefault(x => x.Name.LocalName == k)
                         ?.Value
                         ?.Trim();

                    int? I(string k) =>
                        int.TryParse(S(k), out var v) ? v : null;

                    decimal? D(string k) =>
                        decimal.TryParse(
                            S(k),
                            NumberStyles.Any,
                            CultureInfo.InvariantCulture,
                            out var v
                        ) ? v : null;

                    DateTime? DT(string k)
                    {
                        var raw = S(k);
                        if (string.IsNullOrWhiteSpace(raw)) return null;

                        string[] fmts = { "yyyyMMdd", "yyyyMMddHHmm", "yyyyMMddHHmmss" };
                        return DateTime.TryParseExact(
                            raw,
                            fmts,
                            CultureInfo.InvariantCulture,
                            DateTimeStyles.None,
                            out var d
                        ) ? d : null;
                    }
                    // =================================================

                    var m = new Xml2Model
                    {
                        MA_LK = S("MA_LK"),
                        STT = I("STT"),

                        MA_THUOC = S("MA_THUOC"),
                        MA_PP_CHEBIEN = S("MA_PP_CHEBIEN"),
                        MA_CSKCB_THUOC = S("MA_CSKCB_THUOC"),
                        MA_NHOM = S("MA_NHOM"),

                        TEN_THUOC = S("TEN_THUOC"),
                        DON_VI_TINH = S("DON_VI_TINH"),
                        HAM_LUONG = S("HAM_LUONG"),
                        DUONG_DUNG = S("DUONG_DUNG"),
                        DANG_BAO_CHE = S("DANG_BAO_CHE"),

                        LIEU_DUNG = S("LIEU_DUNG"),
                        CACH_DUNG = S("CACH_DUNG"),
                        SO_DANG_KY = S("SO_DANG_KY"),
                        TT_THAU = S("TT_THAU"),
                        PHAM_VI = S("PHAM_VI"),

                        TYLE_TT_BH = D("TYLE_TT_BH"),
                        SO_LUONG = D("SO_LUONG"),
                        DON_GIA = D("DON_GIA"),

                        THANH_TIEN_BV = D("THANH_TIEN_BV"),
                        THANH_TIEN_BH = D("THANH_TIEN_BH"),

                        T_NGUONKHAC_NSNN = D("T_NGUONKHAC_NSNN"),
                        T_NGUONKHAC_VTNN = D("T_NGUONKHAC_VTNN"),
                        T_NGUONKHAC_VTTN = D("T_NGUONKHAC_VTTN"),
                        T_NGUONKHAC_CL = D("T_NGUONKHAC_CL"),
                        T_NGUONKHAC = D("T_NGUONKHAC"),

                        MUC_HUONG = D("MUC_HUONG"),
                        T_BNTT = D("T_BNTT"),
                        T_BNCCT = D("T_BNCCT"),
                        T_BHTT = D("T_BHTT"),

                        MA_KHOA = S("MA_KHOA"),
                        MA_BAC_SI = S("MA_BAC_SI"),
                        MA_DICH_VU = S("MA_DICH_VU"),

                        NGAY_YL = DT("NGAY_YL"),
                        NGAY_TH_YL = DT("NGAY_TH_YL"),

                        MA_PTTT = S("MA_PTTT"),
                        NGUON_CTRA = S("NGUON_CTRA"),
                        VET_THUONG_TP = S("VET_THUONG_TP"),

                        DU_PHONG = S("DU_PHONG")
                    };
                    result.Add(m);
                }
            }

            return result;
        }
    }
}

File: XMLValidator_v2/src/parsers/parsers/xml2.parser.ts
import { BaseXmlParser } from '../base-xml.parser';
import { Xml2Model } from '../../interfaces/xml-models.interface';

export class Xml2Parser extends BaseXmlParser {
  public get loaiHoSo(): string {
    return 'XML2';
  }

  public parse(xmlContent: string): Xml2Model[] {
    const result: Xml2Model[] = [];

    if (!this.load(xmlContent)) return result;
    if (!this.isRoot('CHITIEU_CHITIET_THUOC')) return result;

    const dsachNodes = this.findChildren(this.root, 'DSACH_CHI_TIET_THUOC');

    for (const dsach of dsachNodes) {
      const itemNodes = this.findChildren(dsach, 'CHI_TIET_THUOC');

      for (const n of itemNodes) {
        const m: Xml2Model = {
          MA_LK: this.getElementText(n, 'MA_LK') || undefined,
          STT: this.getElementInt(n, 'STT') ?? undefined,
          MA_THUOC: this.getElementText(n, 'MA_THUOC') || undefined,
          MA_PP_CHEBIEN: this.getElementText(n, 'MA_PP_CHEBIEN') || undefined,
          MA_CSKCB_THUOC: this.getElementText(n, 'MA_CSKCB_THUOC') || undefined,
          MA_NHOM: this.getElementText(n, 'MA_NHOM') || undefined,
          TEN_THUOC: this.getElementText(n, 'TEN_THUOC') || undefined,
          DON_VI_TINH: this.getElementText(n, 'DON_VI_TINH') || undefined,
          HAM_LUONG: this.getElementText(n, 'HAM_LUONG') || undefined,
          DUONG_DUNG: this.getElementText(n, 'DUONG_DUNG') || undefined,
          DANG_BAO_CHE: this.getElementText(n, 'DANG_BAO_CHE') || undefined,
          LIEU_DUNG: this.getElementText(n, 'LIEU_DUNG') || undefined,
          CACH_DUNG: this.getElementText(n, 'CACH_DUNG') || undefined,
          SO_DANG_KY: this.getElementText(n, 'SO_DANG_KY') || undefined,
          TT_THAU: this.getElementText(n, 'TT_THAU') || undefined,
          PHAM_VI: this.getElementText(n, 'PHAM_VI') || undefined,
          TYLE_TT_BH: this.getElementDecimal(n, 'TYLE_TT_BH') ?? undefined,
          SO_LUONG: this.getElementDecimal(n, 'SO_LUONG') ?? undefined,
          DON_GIA: this.getElementDecimal(n, 'DON_GIA') ?? undefined,
          THANH_TIEN_BV: this.getElementDecimal(n, 'THANH_TIEN_BV') ?? undefined,
          THANH_TIEN_BH: this.getElementDecimal(n, 'THANH_TIEN_BH') ?? undefined,
          T_NGUONKHAC_NSNN: this.getElementDecimal(n, 'T_NGUONKHAC_NSNN') ?? undefined,
          T_NGUONKHAC_VTNN: this.getElementDecimal(n, 'T_NGUONKHAC_VTNN') ?? undefined,
          T_NGUONKHAC_VTTN: this.getElementDecimal(n, 'T_NGUONKHAC_VTTN') ?? undefined,
          T_NGUONKHAC_CL: this.getElementDecimal(n, 'T_NGUONKHAC_CL') ?? undefined,
          T_NGUONKHAC: this.getElementDecimal(n, 'T_NGUONKHAC') ?? undefined,
          MUC_HUONG: this.getElementDecimal(n, 'MUC_HUONG') ?? undefined,
          T_BNTT: this.getElementDecimal(n, 'T_BNTT') ?? undefined,
          T_BNCCT: this.getElementDecimal(n, 'T_BNCCT') ?? undefined,
          T_BHTT: this.getElementDecimal(n, 'T_BHTT') ?? undefined,
          MA_KHOA: this.getElementText(n, 'MA_KHOA') || undefined,
          MA_BAC_SI: this.getElementText(n, 'MA_BAC_SI') || undefined,
          MA_DICH_VU: this.getElementText(n, 'MA_DICH_VU') || undefined,
          NGAY_YL: this.getElementDate(n, 'NGAY_YL') ?? undefined,
          NGAY_TH_YL: this.getElementDate(n, 'NGAY_TH_YL') ?? undefined,
          MA_PTTT: this.getElementText(n, 'MA_PTTT') || undefined,
          NGUON_CTRA: this.getElementText(n, 'NGUON_CTRA') || undefined,
          VET_THUONG_TP: this.getElementText(n, 'VET_THUONG_TP') || undefined,
          DU_PHONG: this.getElementText(n, 'DU_PHONG') || undefined,
        };
        result.push(m);
      }
    }

    return result;
  }
}
